### PR TITLE
Refactor I&A signer modules to reduce API surface

### PR DESCRIPTION
**Pull request by haydenbaker**
## Motivation and Context
- To reduce the amount of protected APIs and code duplication, we can merge the signer modules into the singular http-auth-aws module

## Modifications
- Adds placeholder classes in the http-auth-aws-crt and http-auth-aws-event-stream modules, and uses them in loader runtime checks to make sure the required modules are added
  - This is necessary because these dependencies are now optional deps in the http-auth-aws pom
  - http-auth-aws-crt requires aws-crt in its pom
  - http-auth-aws-event-stream requires eventstream in its pom
  - This allows us to have compile-time typing, while having a negligible impact on the jar size of the main module (http-auth-aws)
- Moves code from other signer modules into http-auth-aws, following the same paths
- Refactor all protected APIs to be completely internal APIs
- Re-formats all of the files based on the team's checkstyle configuration (via IDEA)
- Removes redundant builder interfaces (not needed for internal classes)
- Adds tests to verify behavior when optional modules are not present (simulated)

## Testing
- Run tests, make sure new and existing pass

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+     * exception based on why it failed to load. This should be used in cases where certain dependencies are optional, but the
+     * dependency is used at compile-time for strong typing (i.e. {@link EventStreamV4PayloadSigner}).
+     */
+    public static void requireClass(String classPath, String module) {
+        try {
+            ClassLoaderHelper.loadClass(classPath, false);
+        } catch (ClassNotFoundException e) {
+            LOG.debug(() -> "Cannot find the " + classPath + " class: ", e);
+            String msg = String.format(
+                "Could not load class (%s). You must add a dependency on the '%s' module to enable this functionality: ",
```
Comment (gosar) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java:57]: I'm not sure we want the class names of these empty classes we have as internal implementation detail to be customer visible. Also "this functionality" may be confusing to customer? I like Matt's suggestion `Event stream signing requires that you add a dependency on 'software.amazon.awssdk:http-auth-aws-event-streams' module`. Note, it also includes the group name. You might pass `String feature="Event stream signing"` as additional parameter to requireClass.

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.utils.ClassLoaderHelper;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * Utilities for loading of classes and objects which have optional dependencies, and therefore need to be safely checked at
+ * runtime in order to use.
+ */
+@SdkInternalApi
+public final class LoaderUtil {
```
Comment (gosar) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java:33]: nit: internal so naming doesn't matter too much, but suggest `OptionalDependencyLoaderUtil`

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                module
+            );
+
+            throw new RuntimeException(msg, e);
+        } catch (Exception e) {
+            throw new RuntimeException(String.format("Could not load class (%s): ", classPath), e);
+        }
+    }
+
+    public static AwsV4aHttpSigner getAwsV4aHttpSigner() {
```
Comment (gosar) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java:68]: nit: suggest having this `getDefaultAwsCrtV4aHttpSigner` and returning `DefaultAwsCrtV4aHttpSigner`, similar to `EventStreamV4PayloadSigner getEventStreamV4PayloadSigner`. The place where this method is called knows it is trying to specifically get `DefaultAwsCrtV4aHttpSigner`

**Review by gosar — APPROVED** [review]

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/checksums/Crc32CChecksum.java [hunk 4]
```diff
@@ -44,12 +44,27 @@ public Crc32CChecksum() {
         }
     }
 
+    public static boolean isCrtAvailable() {
```
Comment (zoewangg) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/checksums/Crc32CChecksum.java:47]: `private`?

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/checksums/Crc32Checksum.java [hunk 5]
```diff
@@ -44,6 +44,22 @@ public Crc32Checksum() {
         }
     }
 
+    public static boolean isCrtAvailable() {
```
Comment (zoewangg) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/checksums/Crc32Checksum.java:47]: `private`?

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+@SdkInternalApi
+public final class LoaderUtil {
+    private static final Logger LOG = Logger.loggerFor(LoaderUtil.class);
+
+    private static final String HTTP_AUTH_AWS_CRT_PATH =
+        "software.amazon.awssdk.http.auth.aws.crt.HttpAuthAwsCrt";
+    private static final String HTTP_AUTH_AWS_CRT_MODULE = "http-auth-aws-crt";
+    private static final String HTTP_AUTH_AWS_EVENT_STREAM_PATH =
+        "software.amazon.awssdk.http.auth.aws.eventstream.HttpAuthAwsEventStream";
+    private static final String HTTP_AUTH_AWS_EVENT_STREAM_MODULE = "http-auth-aws-event-stream";
```
Comment (zoewangg) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/LoaderUtil.java:41]: can we specify the group ID as well

`software.amazon.awssdk:http-auth-aws-event-stream`

**Review by zoewangg — APPROVED** [review]

**Pull request merged**